integ/wait_until_search_attributes_optimization_test.go

```diff
@@ -0,0 +1,154 @@
+		WorkflowTimeoutSeconds: 10,
+		IwfWorkerUrl:           "http://localhost:" + testWorkflowServerPort,
+		StartStateId:           ptr.Any(wait_until_search_attributes_optimization.State1),
+		WorkflowStartOptions: &iwfidl.WorkflowStartOptions{
+			WorkflowConfigOverride: config,
+		},
+	}
+	_, httpResp, err := reqStart.WorkflowStartRequest(wfReq).Execute()
+	panicAtHttpError(err, httpResp)
+
+	time.Sleep(time.Second * 5)
+
+	signalValue := iwfidl.EncodedObject{
+		Encoding: iwfidl.PtrString("json"),
+		Data:     iwfidl.PtrString("test"),
+	}
+
+	reqSignal := apiClient.DefaultApi.ApiV1WorkflowSignalPost(context.Background())
+	httpResp, err = reqSignal.WorkflowSignalRequest(iwfidl.WorkflowSignalRequest{
+		WorkflowId:        wfId,
+		SignalChannelName: wait_until_search_attributes_optimization.SignalName,
+		SignalValue:       &signalValue,
+	}).Execute()
+
+	panicAtHttpError(err, httpResp)
+
+	reqWait := apiClient.DefaultApi.ApiV1WorkflowGetWithWaitPost(context.Background())
+	_, httpResp, err = reqWait.WorkflowGetRequest(iwfidl.WorkflowGetRequest{
+		WorkflowId: wfId,
+	}).Execute()
+	panicAtHttpError(err, httpResp)
+
+	// wait for workflow to complete
+	resp, httpResp, err := reqWait.WorkflowGetRequest(iwfidl.WorkflowGetRequest{
+		WorkflowId: wfId,
+	}).Execute()
+	panicAtHttpErrorOrWorkflowUncompleted(err, httpResp, resp)
+
+	api := uclient.GetApiService().(workflowservice.WorkflowServiceClient)
+	reqHistory := &workflowservice.GetWorkflowExecutionHistoryRequest{
+		Namespace: "default",
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wfId,
+		},
+	}
+	eventHistory, err := api.GetWorkflowExecutionHistory(context.Background(), reqHistory)
+	if err != nil {
+		panic("couldn't load eventHistory")
+	}
+
+	var upsertSAEvents []*history.HistoryEvent
+
+	for _, e := range eventHistory.History.Events {
+		if e.EventType == enums.EVENT_TYPE_UPSERT_WORKFLOW_SEARCH_ATTRIBUTES {
+			upsertSAEvents = append(upsertSAEvents, e)
+		}
+	}
+
+	switch mode := config.GetExecutingStateIdMode(); mode {
+	case iwfidl.ENABLED_FOR_ALL:
+		assertions.Equal(6, len(upsertSAEvents))
+		assertions.Equal("[\"S1\"]", historyEventSAs(upsertSAEvents[0]))
+		assertions.Equal("[\"S2\"]", historyEventSAs(upsertSAEvents[1]))
+		assertions.Equal("[\"S3\",\"S4\"]", historyEventSAs(upsertSAEvents[2]))
+		assertions.Equal("[\"S4\"]", historyEventSAs(upsertSAEvents[3]))
+		assertions.Equal("[\"S5\"]", historyEventSAs(upsertSAEvents[4]))
+		assertions.Equal("null", historyEventSAs(upsertSAEvents[5]))
+	case iwfidl.ENABLED_FOR_STATES_WITH_WAIT_UNTIL:
+		assertions.Equal(5, len(upsertSAEvents))
+		assertions.Equal("[\"S1\"]", historyEventSAs(upsertSAEvents[0]))
+		assertions.Equal("[\"S2\"]", historyEventSAs(upsertSAEvents[1]))
+		assertions.Equal("[\"S3\",\"S4\"]", historyEventSAs(upsertSAEvents[2]))
+		assertions.Equal("[\"S4\"]", historyEventSAs(upsertSAEvents[3]))
+		assertions.Equal("null", historyEventSAs(upsertSAEvents[4]))
```

Comment:
I think I tested all the usecases we talked about @longquanzheng:
- S1 loops to itself 5 times
- S3, S4 -> S4 is the one where `slices.Contains` check would fail

Reply:
On the second thought, I'll add one more. S3, S4 -> S4 **-> S4, S5, ->** S5 -> END